### DISCUSSION TITLE
ESP32

### DISCUSSION THREAD

**Discussion by Mulinby** [codewrite/arduino-capacitor#4] [Ideas]
Hello, thank you for the idea. I tried to modify your code from the original article for ESP32:
const int MAX_ADC_VALUE = 4095;
const float R_PULLUP = 45; // According to the datasheet
const float IN_CAP_TO_GND = 2.00; // According to the datasheet
and used other pins:
const int OUT_PIN = 32;
const int IN_PIN = 27;
But I get not quite correct values.
I even soldered a 37pF capacitor between IN_PIN and pin ground:
const float IN_CAP_TO_GND = 39.00
And still the values for if (val> 3700) do not correspond to reality

**Comment by codewrite** [comment]
Hi, Have you got the schematic for the board? One thing to check is if there is a resistor connected to the analog pin. If there is it won't work. Reasons why there might be a resistor is a pullup / pulldown or voltage divider - which I have seen on a lot of the 3.3V boards.

**Reply by Mulinby** [reply]
The resistor is only internal, according to the scheme it is not. But I don't use it.

```
const int OUT_PIN = 32;
const int IN_PIN = 27;
const float IN_CAP_TO_GND  = 70.00;
const float R_PULLUP = 45;
const int MAX_ADC_VALUE = 4095;

void setup()
{
  Serial.begin(115200);
  pinMode(OUT_PIN, OUTPUT);
  pinMode(IN_PIN, OUTPUT);

}

void loop()
{
  Serial.println(calc());
}

String calc()
{
  String value;
  pinMode(IN_PIN, INPUT);
  digitalWrite(OUT_PIN, HIGH);
  int val = analogRead(IN_PIN);
  digitalWrite(OUT_PIN, LOW);
  Serial.println(val);

  if (val < 2000)
  {
    pinMode(IN_PIN, OUTPUT);
    digitalWrite(IN_PIN, LOW);
    float capacitance = (float)val * IN_CAP_TO_GND / (float)(MAX_ADC_VALUE - val);
    value = String(capacitance) + " pF";
  }
  else
  {
    pinMode(IN_PIN, OUTPUT);
    delay(1);
    pinMode(OUT_PIN, INPUT_PULLUP);
    unsigned long u1 = micros();
    unsigned long t;
    int digVal;
    do
    {
      digVal = digitalRead(OUT_PIN);
      unsigned long u2 = micros();
      t = u2 > u1 ? u2 - u1 : u1 - u2;
    } while ((digVal < 1) && (t < 400000L));

    pinMode(OUT_PIN, INPUT);
    val = analogRead(OUT_PIN);
    digitalWrite(IN_PIN, HIGH);
    int dischargeTime = (int)(t / 1000L) * 5;
    delay(dischargeTime);
    pinMode(OUT_PIN, OUTPUT);
    digitalWrite(OUT_PIN, LOW);
    digitalWrite(IN_PIN, LOW);

    float capacitance = -(float)t / R_PULLUP
                        / log(1.0 - (float)val / (float)MAX_ADC_VALUE);

    if (capacitance > 1000.0)
    {
      value = String(capacitance) + " uF";
    }
    else
    {
      value = String(capacitance) + " nF";
    }
  }
  return value;
  while (millis() % 1000 != 0);
}

```

**Reply by codewrite** [reply]
I'm not sure. My only suggestion would be to find the biggest resistor you can find (e.g. 10M ohms) and connect it between Vcc and IN_PIN, and read the ADC value with `int val = analogRead(IN_PIN)`. It should read MAX_ADC_VALUE, i.e. 4095. Then connect the resistor between Gnd and IN_PIN and read the ADC value. It should read 0.
Also, I notice from the docs (https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/adc.html?highlight=adc#analog-to-digital-converter) that GPIO27 is on ADC2 which is used by the WiFi driver. So maybe try swapping IN_PIN and OUT_PIN, i.e. IN_PIN = 32 and OUT_PIN = 27 ?
